src/databricks/labs/ucx/source_code/dependencies.py

```diff
@@ -99,12 +101,33 @@ def _load_source(self, object_info: ObjectInfo) -> str:
             return f.read().decode("utf-8")
 
 
+class DependencyResolver:
+    def __init__(self, whitelist: Whitelist | None = None):
+        self._whitelist = Whitelist() if whitelist is None else whitelist
+
+    def resolve(self, dependency: Dependency) -> Dependency | None:
+        if dependency.type == ObjectType.NOTEBOOK:
+            return dependency
+        compatibility = self._whitelist.compatibility(dependency.path)
+        # TODO attach compatibility to dependency, see https://github.com/databrickslabs/ucx/issues/1382
```

Comment:
Detecting dependencies and detecting compatibility issues are 2 different activities. 
Focusing on the former in this PR.
For the latter, the idea would be to attach compatibility to the dependency, use that to generate Advices, then ignore the dependency (so as not to scan files in the dependency)